EvaluableExpression_client.go

```diff
@@ -0,0 +1,221 @@
+package govaluate
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"time"
+)
+
+// ToClientExpression Returns a string representing this expression in a simplified and client compatible way
+func (expression EvaluableExpression) ToClientExpression(parameters map[string]interface{}) (string, error) {
+
+	var stream *tokenStream
+	var transactions *expressionOutputStream
+	var transaction string
+	var err error
+
+	transactions = new(expressionOutputStream)
+
+	ret := expression.traverseStages(expression.evaluationStages, parameters)
+	stream = newTokenStream(ret)
+	for stream.hasNext() {
+
+		transaction, err = expression.findNextClientString(stream, transactions, parameters)
+		if err != nil {
+			return "", err
+		}
+
+		transactions.add(transaction)
+	}
+
+	return transactions.createString(" "), nil
+}
+
+// Traverse the simplified tree and return the expression tokens in order
+func (expression EvaluableExpression) traverseStages(rootStage *evaluationStage, parameters map[string]interface{}) []ExpressionToken {
+	ret := []ExpressionToken{}
+
+	if rootStage != nil {
+
+		if rootStage.symbol == NOOP {
+			clause := ExpressionToken{
+				Kind: CLAUSE,
+			}
+			ret = append(ret, clause)
+		}
+
+		if rootStage.leftStage == nil && rootStage.rightStage == nil {
+			return []ExpressionToken{rootStage.originalToken}
+		}
+
+		left := expression.traverseStages(rootStage.leftStage, parameters)
+		ret = append(ret, left...)
+
+		if rootStage.originalToken.Value != nil {
+			ret = append(ret, rootStage.originalToken)
+		}
+
+		right := expression.traverseStages(rootStage.rightStage, parameters)
+		ret = append(ret, right...)
+
+		if rootStage.symbol == NOOP {
+			clause := ExpressionToken{
+				Kind: CLAUSE_CLOSE,
+			}
+			ret = append(ret, clause)
+		}
+
+		ret = simplifyTokens(ret, parameters)
+
+	}
+	return ret
+}
+
+func simplifyTokens(tokens []ExpressionToken, parameters map[string]interface{}) []ExpressionToken {
```

Comment:
This function is very dumb and brute-forcey. There are probably much better ways of passing in partial known values and simplifying the expression.